server/src/main/java/org/elasticsearch/index/mapper/ValueFetcher.java

```diff
@@ -35,24 +35,26 @@ public interface ValueFetcher {
      * Note that for array values, the order in which values are returned is undefined and
      * should not be relied on.
      *
-     * @param lookup        a lookup structure over the document's source.
+     * @param source        the document's source.
+     * @param doc           the document id
      * @param ignoredValues a mutable list to collect any ignored values as they were originally presented in source
      * @return a list a standardized field values.
      */
-    List<Object> fetchValues(SourceLookup lookup, List<Object> ignoredValues) throws IOException;
+    List<Object> fetchValues(Source source, int doc, List<Object> ignoredValues) throws IOException;
```

Comment:
This is a great step forward: I wonder long term whether we still want ValueFetcher to be source-centric. There are already implementations that don't retrieve from source, which with this change ignore the Source argument completely. I was planning on addressing some of this as a follow-up to #87267 but I never did.